README

JBIG-KIT lossless image compression library
-------------------------------------------

by Markus Kuhn


The latest release of JBIG-KIT can be downloaded from

  http://www.cl.cam.ac.uk/~mgk25/jbigkit/

JBIG-KIT implements a highly effective data compression algorithm for
bi-level high-resolution images such as fax pages or scanned
documents.

JBIG-KIT provides two variants of a portable library of compression
and decompression functions with a documented interface. You can very
easily include into your image or document processing software. In
addition, JBIG-KIT provides ready-to-use compression and decompression
programs with a simple command line interface (similar to the
converters found in Jef Poskanzer's PBM graphics file conversion
package).

JBIG-KIT implements the specification

  International Standard ISO/IEC 11544:1993 and ITU-T Recommendation
  T.82(1993), "Information technology - Coded representation of picture
  and audio information - progressive bi-level image compression",
  <http://www.itu.int/rec/T-REC-T.82>,

which is commonly referred to as the "JBIG1 standard". JBIG (Joint
Bi-level Image experts Group) is the committee which developed this
international standard for the lossless compression of images using
arithmetic coding. Like the well-known compression algorithms JPEG and
MPEG, JBIG has also been developed and published by the International
Organization for Standardization (ISO) and the International
Telecommunication Union (ITU). See also

  http://www.jpeg.org/jbig/
  http://www.iso.ch/
  http://www.itu.int/

The JBIG compression algorithm offers the following features:

  - Close to state-of-the-art lossless compression ratio for high
    resolution bi-level images.

  - About 1.1 to 1.5 times better compression ratio on typical
    scanned documents compared to G4 fax compression (ITU-T T.6),
    which has been the best compression algorithm for scanned
    documents available prior to JBIG.

  - Up to 30 times better compression of scanned images with dithered
    images compared to G4 fax compression.

  - About 2 times better compression on typical 300 dpi documents
    compared to 'gzip -9' on raw bitmaps.

  - About 3-4 times better compression than GIF on typical 300 dpi
    documents.

  - Even much better competitive compression results on computer
    generated images which are free of scanning distortions.

  - JBIG supports hierarchical "progressive" encoding, that means it is
    possible to encode a low resolution image first, followed by
    resolution enhancement data. This allows, for instance, a document
    browser to display already a good 75 dpi low resolution version of
    an image, while the data necessary to reconstruct the full 300 dpi
    version for laser printer reproduction is still arriving (say
    over a slow network link or mass storage medium).

  - The various resolution layers of a JBIG image in progressive
    encoding mode together require not much more space than a
    normal non-progressive mode encoded image (which JBIG also
    supports).

  - The progressive encoding mode utilizes a quite sophisticated
    resolution reduction algorithm which offers high quality low
    resolution versions that preserve the shape of characters as well
    as the integrity of thin lines and dithered images.

  - JBIG supports multiple bit planes and can this way also be used
    for grayscale and color images, although the main field of
    application is compression of bi-level images, i.e. images with
    only two different pixel values. For grayscale images with up to
    6 bit per pixel, JBIG performs superior to JPEG's lossless
    mode.

JBIG-KIT can be used as free software under the GNU General Public
License. Other license arrangements more suitable for commercial
applications are available as well, please contact the author for
details. JBIG-KIT provides two portable libraries implemented in
ANSI/ISO C for encoding and decoding JBIG data streams, along with
documentation. The first library, jbig.c, implements nearly all of the
options that the JBIG standard provides, but keeps the entire
uncompressed image in memory. The second library, jbig85.c, implements
only the ITU-R T.85 subset of the standard that black/white fax
machines use (single bit per pixel, no "progressive" encoding), and
keeps only three lines of the uncompressed image in memory, making it
particularly attractive for low-memory embedded applications.

The libraries are not intended for 8-bit or 16-bit machine
architectures (e.g., old MS-DOS C compilers). For maximum performance,
a 32-bit processor is required (64-bit systems work too, of course).
On architectures with 16-bit pointer arithmetic, the full-featured
jbig.c library can process only very small images.

Special features of the full-featured jbig.c variant:

  - Fully reentrant multithread-capable design (no global or static
    variables, isolated malloc()/free() calls, etc.)

  - Capable of handling incomplete and growing JBIG data streams in
    order to allow earliest display of low resolution versions

  - Capable of handling several incoming data streams simultaneously
    in one single process and thread

  - Especially designed with applications in mind that want to display
    incoming data as early as possible (e.g., similar to the way in
    which Netscape Navigator handles incoming GIF images)

  - Implements all JBIG features and options including progressive and
    sequential encoding, multiple bit planes, user specified
    resolution reduction and deterministic prediction tables, adaptive
    template changes for optimal performance on half-tone images,
    deterministic prediction, typical prediction in lowest and
    differential layers, various stripe orderings, etc; only the SEQ
    and HITOLO options are currently not supported by the decoder
    (they are normally never required, but could be added later in
    case of user requirements)

  - Suitable for fax applications, satisfies ITU-T T.85 profile

  - Efficient code, optimized utilization of 32-bit processor
    registers

  - Very easy to use documented C library interface

  - Included Gray code conversion routines for efficient encoding
    of grayscale images

  - Ready-to-use pbmtojbg and jbgtopbm converters.

Special features of the light-weight jbig85.c variant:

  - Suitable for low-memory embedded applications

  - Implements only the JBIG1 subset defined in the ITU-T T.85
    profile (single bit plane, no differential layers)

  - Requires only three pixel rows of the uncompressed image to be
    kept in memory

  - Handles all NEWLEN modes of operation required by ITU-T T.85 with
    just a single pass over the data, automatically performing the
    necessary lookahead after the last stripe

  - Codec buffers only a few bytes of arithmetic-codec data and outputs
    resulting bytes or lines as soon as they are available.

I will try to provide free support and maintenance for this software
for the foreseeable future, depending on my available time.

Happy compressing ...

Markus Kuhn

--
Markus Kuhn, Computer Laboratory, University of Cambridge
http://www.cl.cam.ac.uk/~mgk25/ || CB3 0FD, Great Britain


CHANGES

JBIG-KIT revision history
-------------------------


Changes in version 2.1 (2014-04-08)

This is a security-critical bug-fix release that remains API and ABI
backwards compatible to version 2.0. Users who process BIE data from
untrusted sources should upgrade.

  - fixed a buffer-overflow vulnerability in the jbig.c decoder,
    reported by Florian Weimer (Red Hat): CVE-2013-6369

  - fixed ability of corrupted input data to force jbig85.c decoder
    into an end-less loop

  - fixed a bug in the processing of private deterministic-prediction
    tables (DPPRIV=1) in jbig.c decoder

  - fixed integer-type mismatches in printf arguments on 64-bit systems

  - fuzz-testing script added


Changes in version 2.0 (2008-08-30)

Main new features of release 2.0:

  - This release adds a separate lightweight "T.85" version of the
    encoder and decoder library (jbig85.c, jbig85.h). This new T.85
    library is optimized for embedded low-memory applications, such as
    printers and fax machines. It implements only the subset of the
    JBIG standard that is required by the ITU-T T.85 fax profile,
    namely only a single bit plane (P = 1) and no differential layers
    (DL = D = HITOLO = SEQ = ILEAVE = SMID = TPDON = DPON = DPON =
    DPLAST = 0) and some other restrictions (MY = 0, MX < 128).

    The T.85 library requires only three pixel rows of the
    uncompressed image to reside in memory at any time. This 3-line
    buffer has to be allocated by the calling application, therefore
    the T.85 library will not call any heap management functions such
    as malloc() or free() itself. It can fully deal with endless-paper
    fax, where the height of the encoded image is not known initially.

  - There are also two new example applications pbmtojbg85 and
    jbgtopbm85 included that provide command-line access to the new
    T.85 library.

API changes in 2.0 to the original full jbig.c library (compared to 1.6):

  - jbg_dec_getwidth(), jbg_dec_getheight(), jbg_dec_getsize(), and
    jbg_dec_getsize_merged() return now an unsigned long value
    (was: long), and return 0 (was: -1) if no image is available yet

  - jbg_dec_in() return values changed; the least significant bits of
    the error codes now provide more information about what caused the
    error, beyond what the more general error text messages returned by
    jbg_strerror() reveal

  - non-English translations of error messages are now provided
    in a separate PO file (e.g., for use with GNU gettext),
    jbg_strerror() no longer has a language attribute and the
    JBG_EN, JBG_DE_8859_1, JBG_DE_UTF_8 constants no longer exist

Other changes in 2.0 to the original full jbig.c library:

  - the arithmetic codec routines are now located in a separate source
    file (they are now used by both the full library jbig.c and the new
    T.85 library jbig85.c, and might also be of use to other applications)

  - arithmetic decoder behaviour slightly changed to enable the
    lookahead needed to handle T.85-style NEWLEN after the final
    stripe

  - added encoder support for adding comment marker segment

  - added encoder option for using SDRST marker instead of SDNORM
    (only useful to make encoder able to generate a wider range
    of test data)

  - macro JBG_LICENCE can be changed from the default value "GPL" to a
    licence agreement reference code to if the library is used under a
    commercial licence, to clarify under which exact licence the
    library is used in a particular application

Changes in 2.0 to the command-line wrappers for jbig.c:

  - added pbmtojbg options -C and -r, which provide command-line
    access to the previous two extensions

  - slightly improved diagnostic output of jbgtopbm option -d

  - pbmtojbg has a new option -f as a shortcut for all the options
    needed to make the output comply to the basic ITU-T T.85 fax
    profile

  - jbgtopbm now outputs image dimensions in header padded to 10
    digits fixed width, for consistency with jbgtopbm85
    to allow for later in-place update of height due to NEWLEN


Changes in version 1.6 (2004-06-11)

  - various small changes to reduce the risk of 32-bit unsigned
    integer overflows when dealing with extremely large images

  - robuster treatment of L0 = 0xffffffff

  - minor API modification in jbg_enc_options(): parameter l0 changed
    from type long to unsigned long; previous value now remains
    unchanged when l0 == 0 (was: l0 < 0)

  - lots of type casts added such that the C source code is now
    also compilable as C++


Changes in version 1.5 (2003-06-11)

  - fixed two minor memory leaks (special thanks to Davide Pizzolato
    <ing.davide.pizzolato@libero.it> for locating one of these)

  - jbgtopbm does not attempt any more to parse multiple concatenated
    BIEs (options -m must be used now to enable this feature explicitly),
    in order to handle BIEs with data after the last expected SDE gracefully

  - various extensions to improve support of JBIG fax applications
    (ITU-T T.85 profile):

      o support for maximum adaptive template pixel offset increased
        to MX=127 in both encoder and decoder

      o encoder now has a hook for producing BIEs with a NEWLEN marker
        segment and VLENGTH=1, in order to assist in testing decoders
        for T.85 conformance (see also new pbmtojbg option -Y)

      o a new function jbg_newlen() can be used to scan an
        entire in-memory BIE for NEWLEN marker segments and update the
        YD value in the BIE header, which can be applied to BIE data
        before passing it to the decoder for T.85 compliance
        (Background: the incremental design of the JBIG-KIT decoder
        makes it infeasible to look ahead for NEWLEN marker segments
        that occur after the SDE with the last image line, therefore
        this second pass is necessary to handle the old-style fax
        applications permitted by T.85 in which the low-RAM encoder

pbmtools/pbmtojbg.c
/*
 *  pbmtojbg - Portable Bitmap to JBIG converter
 *
 *  Markus Kuhn - http://www.cl.cam.ac.uk/~mgk25/jbigkit/
 */

#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h>
#include <limits.h>
#include "jbig.h"


char *progname;                  /* global pointer to argv[0] */
unsigned long total_length = 0;  /* used for determining output file length */


/*
 * Print usage message and abort
 */
static void usage(void)
{
  fprintf(stderr,
     "PBMtoJBIG converter " JBG_VERSION " -- "
     "creates bi-level image entity (BIE) as output file\n\n"
     "usage: %s [<options>] [<input-file> | -  [<output-file>]]\n\n"
     "options:\n\n", progname);
  fprintf(stderr,
     "  -q\t\tsequential coding, no differential layers (like -d 0)\n"
     "  -x number\tmaximum width of lowest resolution layer (default 640)\n"
     "  -y number\tmaximum height of lowest resolution layer (default 480)\n"
     "  -l number\tlowest layer written to output file (default 0)\n"
     "  -h number\thighest layer written to output file (default max)\n"
     "  -b\t\tuse binary code for multiple bitplanes (default: Gray code)\n"
     "  -d number\ttotal number of differential layers (overrides -x and -y)\n"
     "  -s number\theight of a stripe in layer 0\n");
  fprintf(stderr,
     "  -m number\tmaximum adaptive template pixel horizontal offset (default 8)\n"
     "  -t number\tencode only that many most significant planes\n"
     "  -o number\torder byte value: add 1=SMID, 2=ILEAVE, 4=SEQ, 8=HITOLO\n"
     "\t\t(default 3 = ILEAVE+SMID)\n"
     "  -p number\toptions byte value: add DPON=4, TPBON=8, TPDON=16, LRLTWO=64\n"
     "\t\t(default 28 = DPON+TPBON+TPDON)\n");
  fprintf(stderr,
     "  -C string\tadd the provided string as a comment marker segment\n"
     "  -c\t\tdelay adaptive template changes to first line of next stripe\n"
	  "\t\t(only provided for a conformance test)\n"
     "  -r\t\tterminate each stripe with SDRST marker\n"
	  "\t\t(only intended for decoder testing)\n" );
  fprintf(stderr,
     "  -Y number\tannounce in header initially this larger image height\n"
     "\t\t(only for generating test files with NEWLEN and VLENGTH=1)\n"
     "  -f\t\tchose encoding options for T.85 fax profile complianance\n"
     "  -v\t\tverbose output\n\n");
  exit(1);
}


/*
 * malloc() with exception handler
 */
void *checkedmalloc(size_t n)
{
  void *p;

  if ((p = malloc(n)) == NULL) {
    fprintf(stderr, "Sorry, not enough memory available!\n");
    exit(1);
  }

  return p;
}


/*
 * Read an ASCII integer number from file f and skip any PBM
 * comments which are encountered.
 */
static unsigned long getint(FILE *f)
{
  int c;
  unsigned long i;

  while ((c = getc(f)) != EOF)
    if (c == '#')
      while ((c = getc(f)) != EOF && !(c == 13 || c == 10)) ;
    else if (!isspace(c))
      break;
  if (c == EOF) return 0;
  ungetc(c, f);
  if (fscanf(f, "%lu", &i) != 1) {
    fprintf(stderr, "Unsigned integer value expected!\n");
    exit(1);
  }

  return i;
}


/*
 * Callback procedure which is used by JBIG encoder to deliver the
 * encoded data. It simply sends the bytes to the output file.
 */
static void data_out(unsigned char *start, size_t len, void *file)
{
  fwrite(start, len, 1, (FILE *) file);
  total_length += len;
  return;
}


int main (int argc, char **argv)
{
  FILE *fin = stdin, *fout = stdout;
  const char *fnin = NULL, *fnout = NULL;
  int i, j, c;
  int all_args = 0, files = 0;
  unsigned long x, y;
  unsigned long width, height, max, v;
  unsigned long bpl;
  int bpp, planes, encode_planes = -1;
  size_t bitmap_size;
  char type;
  unsigned char **bitmap, *p, *image;
  struct jbg_enc_state s;
  int verbose = 0, delay_at = 0, reset = 0, use_graycode = 1;
  long mwidth = 640, mheight = 480;
  int dl = -1, dh = -1, d = -1, mx = -1;
  unsigned long l0 = 0, y1 = 0;
  char *comment = NULL;
  int options = JBG_TPDON | JBG_TPBON | JBG_DPON;
  int order = JBG_ILEAVE | JBG_SMID;

  /* parse command line arguments */
  progname = argv[0];
  for (i = 1; i < argc; i++) {
    if (!all_args && argv[i][0] == '-')
      if (argv[i][1] == 0) {
	if (files++) usage();
      } else
	for (j = 1; j > 0 && argv[i][j]; j++)
	  switch(argv[i][j]) {
	  case '-' :
	    all_args = 1;
	    break;
	  case 0 :
	    if (files++) usage();
	    break;
	  case 'v':
	    verbose = 1;
	    break;
	  case 'b':
	    use_graycode = 0;
	    break;
	  case 'c':
	    delay_at = 1;
	    break;
	  case 'r':
	    reset = 1;
	    break;
	  case 'f':
	    d = 0;
	    order = 0;
            options = 8;
	    l0 = 128;
	    encode_planes = 1;
	    mx = 127;
	    break;
	  case 'x':
	    if (++i >= argc) usage();
	    j = -1;
	    mwidth = atol(argv[i]);
	    break;
	  case 'y':
	    if (++i >= argc) usage();
	    j = -1;
	    mheight = atol(argv[i]);
	    break;
	  case 'Y':
	    if (++i >= argc) usage();
	    j = -1;
	    y1 = atol(argv[i]);
	    break;
	  case 'o':
	    if (++i >= argc) usage();
	    j = -1;
	    order = atoi(argv[i]);
	    break;
	  case 'p':
	    if (++i >= argc) usage();
	    j = -1;
	    options = atoi(argv[i]);
	    break;
	  case 'l':
	    if (++i >= argc) usage();
	    j = -1;
	    dl = atoi(argv[i]);
	    break;
	  case 'h':
	    if (++i >= argc) usage();
	    j = -1;
	    dh = atoi(argv[i]);
	    break;
	  case 'q':
	    d = 0;
	    break;
	  case 'd':
	    if (++i >= argc) usage();
	    j = -1;
	    d = atoi(argv[i]);
	    break;
	  case 's':
	    if (++i >= argc) usage();
	    j = -1;
	    l0 = atol(argv[i]);
	    break;
	  case 't':
	    if (++i >= argc) usage();
	    j = -1;
	    encode_planes = atoi(argv[i]);
	    break;
	  case 'm':
	    if (++i >= argc) usage();
	    j = -1;
	    mx = atoi(argv[i]);
	    break;
	  case 'C':
	    if (++i >= argc) usage();
	    j = -1;
	    comment = argv[i];
	    break;
	  default:
	    usage();
	  }
    else
      switch (files++) {
      case 0: fnin  = argv[i]; break;
      case 1: fnout = argv[i]; break;
      default:
	usage();
      }
  }

  if (fnin) {
    fin = fopen(fnin, "rb");
    if (!fin) {
      fprintf(stderr, "Can't open input file '%s", fnin);
      perror("'");
      exit(1);
    }
  } else
    fnin  = "<stdin>";
  if (fnout) {
    fout = fopen(fnout, "wb");
    if (!fout) {
      fprintf(stderr, "Can't open input file '%s", fnout);
      perror("'");
      exit(1);
    }
  } else
    fnout = "<stdout>";

  /* read PBM header */
  while ((c = getc(fin)) != EOF && (isspace(c) || c == '#'))
    if (c == '#')
      while ((c = getc(fin)) != EOF && !(c == 13 || c == 10)) ;
  if (c != 'P') {
    fprintf(stderr, "Input file '%s' does not look like a PBM file!\n", fnin);
    exit(1);
  }
  type = getc(fin);
  width = getint(fin);
  height = getint(fin);
  if (width < 1 || height < 1 ||
      width > 0xffffffff || height > 0xffffffff) {
    fprintf(stderr, "Invalid width or height!\n");
    exit(1);
  }
  if (type == '2' || type == '5' ||
      type == '3' || type == '6')
    max = getint(fin);
  else
    max = 1;
  if (max < 1) {
    fprintf(stderr, "Invalid maxval!\n");
    exit(1);
  }
  for (planes = 0, v = max; v; planes++, v >>= 1) ;
  bpp = (planes + 7) / 8;
  if (encode_planes < 1 || encode_planes > planes)
    encode_planes = planes;
  fgetc(fin);    /* skip line feed */

  /* read PBM image data */
  bpl = jbg_ceil_half(width, 3);     /* bytes per line */
  if (ULONG_MAX / height < bpl) {
    fprintf(stderr, "Image too large!\n");
    exit(1);
  }
  bitmap_size = bpl * (size_t) height;
  bitmap = (unsigned char **) checkedmalloc(sizeof(unsigned char *) *
					    encode_planes);
  for (i = 0; i < encode_planes; i++)
    bitmap[i] = (unsigned char *) checkedmalloc(bitmap_size);
  switch (type) {
  case '1':
    /* PBM text format */
    p = bitmap[0];
    for (y = 0; y < height; y++)
      for (x = 0; x <= ((width-1) | 7); x++) {
	*p <<= 1;
	if (x < width)
	  *p |= getint(fin) & 1;
	if ((x & 7) == 7)
	  ++p;
      }
    break;
  case '4':
    /* PBM raw binary format */
    fread(bitmap[0], bitmap_size, 1, fin);
    break;
  case '2':
  case '5':
    /* PGM */
    if ((ULONG_MAX / bpp) / height < width) {
      fprintf(stderr, "Image too large!\n");
      exit(1);
    }
    image = (unsigned char *) checkedmalloc(width * height * bpp);
    if (type == '2') {
      for (x = 0; x < width * height; x++) {
	v = getint(fin);
	for (j = 0; j < bpp; j++)
	  image[x * bpp + (bpp - 1) - j] = v >> (j * 8);
      }
    } else
      fread(image, width * height, bpp, fin);
    jbg_split_planes(width, height, planes, encode_planes, image, bitmap,
		     use_graycode);
    free(image);
    break;
  default:
    fprintf(stderr, "Unsupported PBM type P%c!\n", type);
    exit(1);
  }
  if (ferror(fin)) {
    fprintf(stderr, "Problem while reading input file '%s", fnin);
    perror("'");
    exit(1);
  }
  if (feof(fin)) {
    fprintf(stderr, "Unexpected end of input file '%s'!\n", fnin);
    exit(1);
  }

  /* Test for valid parameters */
  if (width < 1 || height < 1) {
    fprintf(stderr, "Image dimensions must be positive!\n");
    exit(1);
  }
  if (encode_planes < 1 || encode_planes > 255) {
    fprintf(stderr, "Number of planes must be in range 1-255!\n");
    exit(1);
  }

  /* Test the final byte in each image line for correct zero padding */
  if ((width & 7) && type == '4') {
    for (y = 0; y < height; y++)
      if (bitmap[0][y * bpl + bpl - 1] & ((1 << (8 - (width & 7))) - 1)) {
	fprintf(stderr, "Warning: No zero padding in last byte (0x%02x) of "
		"line %lu!\n", bitmap[0][y * bpl + bpl - 1], y + 1);
	break;
      }
  }

  /* Apply JBIG algorithm and write BIE to output file */

  /* initialize parameter struct for JBIG encoder*/
  jbg_enc_init(&s, width, height, encode_planes, bitmap, data_out, fout);

  /* Select number of resolution layers either directly or based
   * on a given maximum size for the lowest resolution layer */
   if (d >= 0)
    jbg_enc_layers(&s, d);
  else
    jbg_enc_lrlmax(&s, mwidth, mheight);

  /* Specify a few other options (each is ignored if negative) */
  if (delay_at)
    options |= JBG_DELAY_AT;
  if (reset)
    options |= JBG_SDRST;
  if (comment) {
    s.comment_len = strlen(comment);
    s.comment = (unsigned char *) comment;
  }
  if (y1)
    s.yd1 = y1;
  jbg_enc_lrange(&s, dl, dh);
  jbg_enc_options(&s, order, options, l0, mx, -1);

  /* now encode everything and send it to data_out() */
  jbg_enc_out(&s);

  /* give encoder a chance to free its temporary data structures */
  jbg_enc_free(&s);

  /* check for file errors and close fout */
  if (ferror(fout) || fclose(fout)) {
    fprintf(stderr, "Problem while writing output file '%s", fnout);
    perror("'");
    exit(1);
  }

  /* In case the user wants to know all the gory details ... */
  if (verbose) {
    fprintf(stderr, "Information about the created JBIG bi-level image entity "
	    "(BIE):\n\n");
    fprintf(stderr, "              input image size: %lu x %lu pixel\n",
	    s.xd, s.yd);
    fprintf(stderr, "                    bit planes: %d\n", s.planes);
    if (s.planes > 1)
      fprintf(stderr, "                      encoding: %s code, MSB first\n",
	      use_graycode ? "Gray" : "binary");
    fprintf(stderr, "                       stripes: %lu\n", s.stripes);
    fprintf(stderr, "   lines per stripe in layer 0: %lu\n", s.l0);
    fprintf(stderr, "  total number of diff. layers: %d\n", s.d);
    fprintf(stderr, "           lowest layer in BIE: %d\n", s.dl);
    fprintf(stderr, "          highest layer in BIE: %d\n", s.dh);
    fprintf(stderr, "             lowest layer size: %lu x %lu pixel\n",
	    jbg_ceil_half(s.xd, s.d - s.dl), jbg_ceil_half(s.yd, s.d - s.dl));
    fprintf(stderr, "            highest layer size: %lu x %lu pixel\n",
	    jbg_ceil_half(s.xd, s.d - s.dh), jbg_ceil_half(s.yd, s.d - s.dh));
    fprintf(stderr, "                   option bits:%s%s%s%s%s%s%s\n",
	    s.options & JBG_LRLTWO  ? " LRLTWO" : "",
	    s.options & JBG_VLENGTH ? " VLENGTH" : "",
	    s.options & JBG_TPDON   ? " TPDON" : "",
	    s.options & JBG_TPBON   ? " TPBON" : "",
	    s.options & JBG_DPON    ? " DPON" : "",
	    s.options & JBG_DPPRIV  ? " DPPRIV" : "",
	    s.options & JBG_DPLAST  ? " DPLAST" : "");
    fprintf(stderr, "                    order bits:%s%s%s%s\n",
	    s.order & JBG_HITOLO ? " HITOLO" : "",
	    s.order & JBG_SEQ    ? " SEQ" : "",
	    s.order & JBG_ILEAVE ? " ILEAVE" : "",
	    s.order & JBG_SMID   ? " SMID" : "");
    fprintf(stderr, "           AT maximum x-offset: %d\n"
	    "           AT maximum y-offset: %d\n", s.mx, s.my);
    fprintf(stderr, "         length of output file: %lu byte\n\n",
	    total_length);
  }

  return 0;
}
